doc: Clarify scope of eviction protection of outbound block-relay peers

ariard opened the pull request:
Block-relay-only peers were introduced by #15759. According to its
author, it was intented to make them only immune to outbound peer
rotation-based eviction and not from all eviction as modified comment
leans to think of.

Clearly indicate that outbound block-relay peers aren't protected
from eviction by the bad/lagging chain logic.

Fix #19863

Commit:
[doc] Clarify scope of eviction protection of outbound block-relay peers

Block-relay-only peers were introduced by #15759. According to its
author, it was intented to make them only immune to outbound peer
rotation-based eviction and not from all eviction as modified comment
leans to think of.

Clearly indicate that outbound block-relay peers aren't protected
from eviction by the bad/lagging chain logic.

sdaftuar commented on src/net_processing.cpp:
```diff
@@ -2021,8 +2021,8 @@ static void ProcessHeadersMessage(CNode& pfrom, CConnman& connman, ChainstateMan
-            // Note that block-relay-only peers are already implicitly protected, so we
-            // only consider setting m_protect for the full-relay peers.
+            // Note that outbound block-relay peers are excluded from this protection and
+            // thus passive of eviction by bad/lagging chain logic.
```
How about:

```
// Note that outbound block-relay peers are excluded from this protection, and thus always subject to eviction under the bad/lagging chain logic.
```

sdaftuar replied:
Also, you may want to consider adding a comment that explains how `m_protect` is used by stale-tip-based eviction as well.  In particular, the stale-tip -> extra full outbound peer logic could cause an existing full outbound to be evicted, but peers with `m_protect` set will not be chosen for eviction.

ariard replied:
Thanks for the suggestion. Also added a commit on scope of m_protect w.r.t to both eviction logics.

sdaftuar reviewed (commented):
The comments above the definition of `ChainSyncTimeoutState` in `net_processing.cpp` should also be updated.

Commit:
[doc] Clarify semantic of peer's m_protect w.r.t to outbound eviction logics

The field m_protect is used to protect from eviction both by bad/lagging
chain and extra outbound peers logics. Outbound block-relay peers are
always excluded from this protection.

ariard commented:
Updated at 6ce273b. I guess we could split up `m_protect` or clarify `ChainSyncTimeout` name but I don't want to interfere with actual refactoring of `CNodeState` (#19398). It would be easier to reorganize structures after completion of this project.

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -2021,8 +2021,8 @@ static void ProcessHeadersMessage(CNode& pfrom, CConnman& connman, ChainstateMan
             // it from the bad/lagging chain logic.
-            // Note that block-relay-only peers are already implicitly protected, so we
-            // only consider setting m_protect for the full-relay peers.
+            // Note that outbound block-relay peers are excluded from this protection, and
```
This code is in the conditional `&& pfrom.IsFullOutboundConn()`.

Why comments about `block-relay-only` connections belong in a block about `full conns` at all?

naumenkogs replied:
I would probably prefer to have a good big comment *before* the following line:
```
         if (!pfrom.fDisconnect && pfrom.IsFullOutboundConn() && nodestate->pindexBestKnownBlock != nullptr) {
```

ariard replied:
Moved comment above with new referral to ChainSyncTimeout.

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -334,6 +339,9 @@ struct CNodeState {
       *     reached, disconnect.
+      *
+      * EXTRA_PEER_CHECK_INTERVAL: after each interval, if we have too many outbound peers,
+      * drop to the outbound one that least recently announced us a new block.
```
`* drop to the outbound one that least recently announced us a new block.` -> 
`* drop the outbound one that least recently announced us a new block.` ("to" does not belong here)

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -321,10 +321,15 @@ struct CNodeState {
+      * Any peer protected (m_protect = true) is not chosen for eviction. Block-relay
+      * peers are always excluded from protection. A peer is marked as protected under
+      * the condition of receiving at least a valid connecting header, having a better
+      * chain than us and not having reached yet MAX_OUTBOUND_PEERS_TO_PROTECT_FROM_DISCONNECT.
```
`and not having reached yet MAX_OUTBOUND_PEERS_TO_PROTECT_FROM_DISCONNECT` makes it sound like *the peer* is not having reached, not us. I suggest this:

A peer is marked as protected if
      - it gave us a valid connecting header
      - it has a better chain than we have
      - we haven't reached MAX_OUTBOUND_PEERS_TO_PROTECT_FROM_DISCONNECT yet

ariard commented:
Thanks @naumenkogs, took your last suggestion at b55f979

naumenkogs commented:
ACK b55f9790194746ce0180fc9fc5e085bdb118432c

jonatack commented on src/net_processing.cpp:
```diff
@@ -321,10 +321,17 @@ struct CNodeState {
+      *
+      * Both are only in effect for outbound, non-manual, non-protected connections.
+      * Any peer protected (m_protect = true) is not chosen for eviction. A peer is
+      * marked as protected if :
```
(drive-by review on the GitHub app from phone at lunch, feel free to ignore, will review properly):

perhaps clarity s/if/if all of these are true/

jonatack reviewed (commented):
Concept ACK

jonatack commented on src/net_processing.cpp:
```diff
@@ -334,6 +341,9 @@ struct CNodeState {
       *     reached, disconnect.
+      *
+      * EXTRA_PEER_CHECK_INTERVAL: after each interval, if we have too many outbound peers,
+      * drop the outbound one that least recently announced us a new block.
```
b55f9790 suggestions
```diff
-    /** State used to enforce CHAIN_SYNC_TIMEOUT and EXTRA_PEER_CHECK_INTERVAL logics.
+    /** State used to enforce CHAIN_SYNC_TIMEOUT and EXTRA_PEER_CHECK_INTERVAL logic.
       *
       * Both are only in effect for outbound, non-manual, non-protected connections.
       * Any peer protected (m_protect = true) is not chosen for eviction. A peer is
-      * marked as protected if :
+      * marked as protected if all of these are true:
       *   - its connection type is IsBlockOnlyConn() == false
       *   - it gave us a valid connecting header
-      *   - it has a better chain than we have
       *   - we haven't reached MAX_OUTBOUND_PEERS_TO_PROTECT_FROM_DISCONNECT yet
+      *   - it has a better chain than we have
+      *   - it is not already marked as protected
       *
@@ -343,7 +344,7 @@ struct CNodeState {
       * EXTRA_PEER_CHECK_INTERVAL: after each interval, if we have too many outbound peers,
-      * drop the outbound one that least recently announced us a new block.
+      * drop the outbound one that least-recently announced us a new block.
       */
```

ariard replied:
I'm going to leave it as it is. It's a bit confusing saying that something is "marked as protected if it is not already marked as protected".

Note, actually CHAIN_SYNC_TIMEOUT and EXTRA_PEER_CHECK_INTERVAL are two different logics, thus the plural.

jonatack replied:
I'm aware. Logic is most often used as an uncountable noun.

While rewriting this, I think it would be good to clarify that the list of conditions is "if all of these" and not "if any of these".

ariard replied:
Thanks for the noun suggestion. Took it, also the re-ordering. But not the "it is not already marked as protected" it's a tautology IMO (it would false future triggering of protection if we don't due to g_outbound_peers_with_protect_from_disconnect is incremented in same conditional, but technically the state transition is defined without)

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -321,10 +321,17 @@ struct CNodeState {
+    /** State used to enforce CHAIN_SYNC_TIMEOUT and EXTRA_PEER_CHECK_INTERVAL logic.
+      *
+      * Both are only in effect for outbound, non-manual, non-protected connections.
+      * Any peer protected (m_protect = true) is not chosen for eviction. A peer is
```
I think this explanation actually belongs to the declaration of `m_protect` better, but I don't care enough to ask for this... Consider this if you gonna update it.

jonatack commented:
ACK d76925478efd35e6fd835370639f2139b28381e4

Thanks for updating @ariard.

amitiuttarwar commented on src/net_processing.cpp:
```diff
@@ -321,10 +321,17 @@ struct CNodeState {
+      *   - its connection type is IsBlockOnlyConn() == false
+      *   - it gave us a valid connecting header
+      *   - we haven't reached MAX_OUTBOUND_PEERS_TO_PROTECT_FROM_DISCONNECT yet
+      *   - it has a better chain than we have
```
I don't think this is quite right? 

the relevant part of the check in `ProcessHeadersMessage` checks `nodestate->pindexBestKnownBlock->nChainWork >= ::ChainActive().Tip()->nChainWork`, aka would set `m_protect` to true if they have the same chain tip as we do. 

I think saying the peer "has a better chain" is misleading, because then I'd expect us to usually not have protected peers. Based on the logic I see, I'd expect us to usually have `MAX_OUTBOUND_PEERS_TO_PROTECT_FROM_DISCONNECT`.

what do you think?
